gateway/src/node/lp_listener/handler.rs

```diff
@@ -0,0 +1,996 @@
+                        e
+                    ))
+                })?;
+
+                // Extract salt for PSK derivation
+                let salt = hello_data.salt;
+
+                Ok((client_pubkey, client_ed25519_pubkey, salt))
+            }
+            other => Err(GatewayError::LpProtocolError(format!(
+                "Expected ClientHello, got {}",
+                other
+            ))),
+        }
+    }
+
+    /// Receive registration request after handshake
+    async fn receive_registration_request(
+        &mut self,
+        session: &LpSession,
+    ) -> Result<LpRegistrationRequest, GatewayError> {
+        // Read LP packet containing the registration request
+        let packet = self.receive_lp_packet().await?;
+
+        // Verify it's from the correct session
+        if packet.header().session_id != session.id() {
+            return Err(GatewayError::LpProtocolError(format!(
+                "Session ID mismatch: expected {}, got {}",
+                session.id(),
+                packet.header().session_id
+            )));
+        }
+
+        // Decrypt the packet payload using the established session
+        let decrypted_bytes = session.decrypt_data(packet.message()).map_err(|e| {
+            GatewayError::LpProtocolError(format!("Failed to decrypt registration request: {}", e))
+        })?;
+
+        // Deserialize the decrypted bytes into LpRegistrationRequest
+        bincode::deserialize(&decrypted_bytes).map_err(|e| {
+            GatewayError::LpProtocolError(format!(
+                "Failed to deserialize registration request: {}",
+                e
+            ))
+        })
+    }
+
+    /// Send registration response after processing
+    async fn send_registration_response(
+        &mut self,
+        session: &LpSession,
+        response: LpRegistrationResponse,
+    ) -> Result<(), GatewayError> {
+        // Serialize response
+        let data = bincode::serialize(&response).map_err(|e| {
+            GatewayError::LpProtocolError(format!("Failed to serialize response: {}", e))
+        })?;
+
+        // Encrypt data first (this increments Noise internal counter)
+        let encrypted_message = session
+            .encrypt_data(&data)
+            .map_err(|e| GatewayError::LpProtocolError(format!("Failed to encrypt data: {}", e)))?;
+
+        // Create LP packet with encrypted message (this increments LP protocol counter)
+        let packet = session.next_packet(encrypted_message).map_err(|e| {
+            GatewayError::LpProtocolError(format!("Failed to create packet: {}", e))
+        })?;
+
+        // Send the packet
+        self.send_lp_packet(&packet).await
+    }
+
+    /// Receive an LP packet from the stream with proper length-prefixed framing
+    async fn receive_lp_packet(&mut self) -> Result<LpPacket, GatewayError> {
+        use nym_lp::codec::parse_lp_packet;
+
+        // Read 4-byte length prefix (u32 big-endian)
+        let mut len_buf = [0u8; 4];
```

Comment:
tihs length prefixing looks like something that should have been handled by a codec and not dealt with manually

Reply:
The issue is that codec is designed for datagram parsing, and not stream parsing, so this is a sort of adapter where we convert the stream to what the codec wants, and the use it